src/active_item_cache.h

```diff
@@ -9,32 +9,52 @@
 #include "enums.h"
 #include "item.h"
 #include "item_stack.h"
+#include "safe_reference.h"
 
 // A struct used to uniquely identify an item within a submap or vehicle.
 struct item_reference {
     point location;
-    item_stack::iterator item_iterator;
-    // Do not access this from outside this module, it is only used as an ID for active_item_set.
-    item *item_id;
+    safe_reference<item> safe_ref;
 };
 
 class active_item_cache
 {
     private:
         std::unordered_map<int, std::list<item_reference>> active_items;
-        // Cache for fast lookup when we're iterating over the active items to verify the item is present.
-        // Key is item_id, value is whether it was returned in the last call to get
-        std::unordered_map<item *, bool> active_item_set;
 
     public:
-        void remove( item_stack::iterator it, point location );
-        void add( item_stack::iterator it, point location );
-        bool has( item_stack::iterator it, point ) const;
-        // Use this one if there's a chance that the item being referenced has been invalidated.
-        bool has( const item_reference &itm ) const;
+        /**
+         * Removes the item if it is in the cache. Does nothing if the item is not in the cache.
+         * Relies on the fact that item::processing_speed() is a constant.
+         * Also removes any items that have been destroyed in the list containing it
+         */
+        void remove( const item *it );
+
+        /**
+         * Adds the reference to the cache. Does nothing if the reference is already in the cache.
+         * Relies on the fact that item::processing_speed() is a constant.
+         */
+        void add( safe_reference<item> it, point location );
```

Comment:
I feel this would be better just to take an `item&` or `item*`; the fact that it's using a `safe_reference` is kind of an implementation detail.

Reply:
Done